### DISCUSSION TITLE
FIPS provider can't be loaded programmatically in OpenSSL 3.1.4.

### DISCUSSION THREAD

**Discussion by Syllinia** [openssl/openssl#25173] [Q&A] [answered]
In my system, I followed "Making all applications use the FIPS module by default" in https://docs.openssl.org/3.0/man7/fips_module/#description. In an initialization process init_A I firstly create the related openssl.cnf and execute fipsinstall.

system("openssl fipsinstall -out /usr/ssl/fipsmodule.cnf -module /lib/ossl-modules/fips.so &> /var/log/gui_upload/fips_install");

Then I'd like to load fips module programmatically. However, at the beginning, fipsmodule.cnf contains a line "activate = 1". And the fips module loading fails.

Then I comment the line "activate = 1" and then uncomment it. But fips module loading still fails.
`static void fips_ssl_providers_load(void)
{
        int fips_load_flag = 0;

        // Programmatically load fips temporarily for process smit.
        int ret = system("sed -i '/^activate = 1/s/^/#/' /usr/ssl/fipsmodule.cnf");
        _CHECK_SYS_RET(ret);

        fips_provider = OSSL_PROVIDER_load(NULL, "fips");
        if (fips_provider == NULL) {
                message("Failed to load FIPS provider\n");
                OSSL_FIPS_DEBUG("Failed to load FIPS provider\n");
                OSSL_FIPS_PRINT_LIB_ERR("");
        } else {
                fips_load_flag = 1;
        }

        // Restore /usr/ssl/fipsmodule.cnf to the form of "making all app use FIPS module by default".
        ret = system("sed -i '/^#activate = 1/s/^#//' /usr/ssl/fipsmodule.cnf");
        _CHECK_SYS_RET(ret);

        if (fips_load_flag) {
                base_provider = OSSL_PROVIDER_load(NULL, "base");
                if (base_provider == NULL) {
                        OSSL_PROVIDER_unload(fips_provider);

                        message("Failed to load base provider\n");
                        OSSL_FIPS_DEBUG("Failed to load base provider\n");
                        OSSL_FIPS_PRINT_LIB_ERR("");
                        return;
                }

                EVP_set_default_properties(NULL, "fips=yes");
        }
}
`
![image](https://github.com/user-attachments/assets/e71f3c6e-c21c-4ad4-9bfe-fd1fc4a98de2)
![image](https://github.com/user-attachments/assets/ce2a3067-e17e-4601-ab04-ae01e9f00bb1)
![image](https://github.com/user-attachments/assets/5128340a-8560-40a1-b60e-84017ffb6bf3)

Other processes started afterward are able to load fips and base providers.
I checked the related error but didn't figure out why it failed to set st->module_checksum_data.

**Comment by kulkarniamit** [comment]
Two key settings might be missing:

* `openssl.cnf` should have a `.include` directive pointing to the full path of `fipsmodule.cnf`
* `openssl.cnf` should contain a reference to the fips section name inside `fipsmodule.cnf` within `[provider_sect]`

You can get the path of your `openssl.cnf` using `$ <install_location>/openssl version -d`.

### Sample setup
```
$ grep "^\.include"  <install_path>/openssl.cnf
.include /tmp/fipsmodule.cnf

$ grep "fips ="  <install_path>//openssl.cnf
fips = fips_sect

$ grep "activate"  /tmp/fipsmodule.cnf
#activate = 1
```

### Failure scenario with incomplete setup
```
$ grep "fips =" <install_path>/openssl.cnf
#fips = fips_sect

$ ./main
FIPS provider load failed
40E758F4C47F0000:error:1C8000D5:Provider routines:SELF_TEST_post:missing config data:providers/fips/self_test.c:290:
40E758F4C47F0000:error:1C8000E0:Provider routines:ossl_set_error_state:fips module entering error state:providers/fips/self_test.c:388:
40E758F4C47F0000:error:1C8000D8:Provider routines:OSSL_provider_init_int:self test post failure:providers/fips/fipsprov.c:707:
40E758F4C47F0000:error:078C0105:common libcrypto routines:provider_init:init fail:crypto/provider_core.c:932:name=fips
```

### Success with complete setup
```
$ ./main
FIPS provider load succeeded
```

### Sample program
```c
#include <stdio.h>
#include <stdlib.h>
#include <openssl/provider.h>
#include <openssl/err.h>
#include <openssl/crypto.h>

int main() {
    OSSL_PROVIDER *fips_provider = NULL;
    fips_provider = OSSL_PROVIDER_load(NULL, "fips");
    fprintf(stdout, "FIPS provider load %s\n",
        (fips_provider == NULL ||
        !OSSL_PROVIDER_available(NULL, "fips")) ? "failed" : "succeeded");
    ERR_print_errors_fp(stderr);
    OSSL_PROVIDER_unload(fips_provider);
    exit(EXIT_SUCCESS);
}
```

**Comment by Syllinia** [comment]
openssl.cnf and fipsmodule.cnf are both good. Other processes started later can load fips and base providers successfully.

![image](https://github.com/user-attachments/assets/0f02dc10-9a4e-4aab-8860-3fe1202287d9)
![image](https://github.com/user-attachments/assets/a5159ba5-2366-4b59-9bbf-b16bf155ebce)

I'd like to use "Making all applications use the FIPS module by default" in https://docs.openssl.org/3.0/man7/fips_module/#description. And the initialization process does the preparation work. But I still want to load fips & base providers in that initialization process.

**Reply by kulkarniamit** [reply, marked as answer]
It would be interesting to try the following two commands

1. Are you able to load automatically activated FIPS provider using openssl command? _If yes_, it would mean FIPS setup is correct and we can take a closer look at the code that's trying to load FIPS module. _If not_, there's an issue with basic openssl fips setup that needs to be fixed before checking the program.
    ```
    $ grep "^\.include"  <install_path>/openssl.cnf
    .include /tmp/fipsmodule.cnf
    
    $ grep "fips =\|base ="  <install_path>/openssl.cnf
    fips = fips_sect
    base = base_sect
    
    $ grep "activate"  /tmp/fipsmodule.cnf
    activate = 1
    
    $ OPENSSL_MODULES=<install_path>/lib64/ossl-modules \
    OPENSSL_CONF=<install_path>/openssl.cnf <install_path>/bin/openssl list -providers
    Providers:
      base
        name: OpenSSL Base Provider
        version: 3.0.9
        status: active
      fips
        name: OpenSSL FIPS Provider
        version: 3.0.9
        status: active
    ```
    * Replace `OPENSSL_MODULES` with path to your fips.so directory
    * Replace `OPENSSL_CONF` with path to your `openssl.cnf` that has included full path to `fipsmodule.cnf`
  
2. Run your program with `$ strace -e open,openat -o /tmp/strace.log.0 <path_to_your_executable>` and check your `strace.log.0` to see if your expected `openssl.cnf`, `fipsmodule.cnf`, `fips.so` are being loaded from the correct path. If there's any unexpected path for any of those files, that could result in failure to load fips provider.

**Reply by Syllinia** [reply]
Sorry for the late response.

1.The openssl works well.
![image](https://github.com/user-attachments/assets/fe994c29-1c2e-44e1-8e38-656e369afbc9)

2.The initialization process is like init in Linux system. So it's difficult to trace it by strace.


Currently, other FIPS related features work well. So I won't load fips & base providers in the initialization process. Thanks so much for your help.
